### PR TITLE
HSEARCH-3122 Restore support for configuring "strict" mapping in Elasticsearch

### PR DESCRIPTION
**Pull request by fax4ever**
https://hibernate.atlassian.net/browse/HSEARCH-3122

~~Draft:~~
~~1. We need more tests: invalid name && propagation on nested object mapping~~
~~2. Documentation~~

### CODE REVIEW

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/logging/impl/Log.java [hunk 1]
```diff
@@ -652,4 +652,8 @@ SearchException differentImplementationClassForQueryElement(@FormatWith(ClassFor
 	SearchTimeoutException clientSideTimedOut(@FormatWith(DurationInSecondsAndFractionsFormatter.class) Duration timeout,
 			String query);
 
+	@Message(id = ID_OFFSET_3 + 121, value = "Invalid dynamic type: '%1$s'."
+			+ " Valid statuses are: %2$s.")
```
Comment (yrodiere) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/logging/impl/Log.java:656]:
```suggestion
			+ " Valid values are: %2$s.")
```

File: documentation/src/main/asciidoc/reference/backend-elasticsearch.asciidoc [hunk 2]
```diff
@@ -612,6 +612,33 @@ contains the actual index name (e.g. `myindex-000001`), not the alias (e.g. `myi
 If indexes do not follow the default naming scheme `<hibernateSearchIndexName>-<6 digits>`,
 a custom <<backend-elasticsearch-indexlayout,index layout>> must be configured.
 
+=== Dynamic types mapping
+
+By default, Hibernate Search set the Elasticsearch mapping dynamic properties to `strict`.
+See dynamic property mapping on
+https://www.elastic.co/guide/en/elasticsearch/reference/current/dynamic.html[Elasticsearch documentation].
+Which means that Elasticsearch server will throw an exception
+if we should try to work on fields that do not belong to the schema defined at bootstrap time.
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/reference/backend-elasticsearch.asciidoc:621]:
```suggestion
By default, Hibernate Search sets the link:{elasticsearchDocUrl}/dynamic.html[`dynamic`] property in Elasticsearch index mappings to `strict`.
This means that attempting to index documents with fields that are not present in the mapping will lead to an indexing failure.
```

File: documentation/src/main/asciidoc/reference/backend-elasticsearch.asciidoc [hunk 3]
```diff
@@ -612,6 +612,33 @@ contains the actual index name (e.g. `myindex-000001`), not the alias (e.g. `myi
...
+hibernate.search.backend.dynamic_mapping = strict (default)
+# To configure a specific index:
+hibernate.search.backend.indexes.<index name>.dynamic_mapping = strict (default)
+----
+
+We said that Hibernate Search **usually** works on declared schema fields. More precisely, it always does
+if no <<mapper-orm-bridge-index-field-dsl-dynamic>> are defined.
+In the case of the root object or a nested object containing some dynamic fields,
+the mapping dynamic properties will be always set to `true` instead,
+regardless the value of dynamic_mapping property.
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/reference/backend-elasticsearch.asciidoc:640]:
```suggestion
When field templates are defined, `dynamic` will be forced to `true`, in order to allow for dynamic fields. In that case, the value of the `dynamic_mapping` property is ignored.
```

**Review by yrodiere — CHANGES_REQUESTED** [review]
Thanks! Just a few comments.

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/document/model/dsl/impl/AbstractElasticsearchIndexSchemaObjectNodeBuilder.java [hunk 4]
```diff
@@ -12,7 +12,7 @@
 import java.util.Map;
 
 import org.hibernate.search.backend.elasticsearch.document.model.impl.AbstractElasticsearchIndexSchemaFieldNode;
-import org.hibernate.search.backend.elasticsearch.lowlevel.index.mapping.impl.DynamicType;
+import org.hibernate.search.backend.elasticsearch.index.DynamicType;
```
[the remark this answers is not in the document]
Comment (fax4ever) [reply backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/document/model/dsl/impl/AbstractElasticsearchIndexSchemaObjectNodeBuilder.java]: Yes move a `Gson` annotated object to an API wasn't a good idea. Fixed!

**Comment by fax4ever** [comment]
We are changing an Elasticsearch parameter, so, even if maybe we shouldn't have issue with the older ES versions, here is a full build: https://ci.hibernate.org/job/hibernate-search/view/change-requests/job/PR-2351/9/.

**Comment by yrodiere** [comment]
Merged, thanks!

**Pull request merged**